### ISSUE TITLE
Freeze when setting pyplot linewidth to a close-to-zero float

### ISSUE THREAD

**Issue by fk3** [matplotlib/matplotlib#14498]
### Bug report

**Bug summary**

When setting the linewidth to a value close to zero (like 1e-15) and having a dashed linestyle in a pyplot, the application freezes.

**Code for reproduction**

```python
from matplotlib import pyplot as plt

x = [0,1]
y = [0,1]

p = plt.plot(x,y, ls = "--", lw = 0.1, ms = 0.1, marker = "o", alpha = 1.0)

# ok
# plt.setp(p, lw = 1e-3)

# freeze
plt.setp(p, lw = 1e-15)

plt.show()
```

**Actual outcome**

The window opens but no plot (not even axis) is shown, instead the window is not responding.

**Expected outcome**

Primary Expectation: The application should not freeze.
Secondary Expectation: Show either no line at all or as thin as displayable.

**Matplotlib version**
  * Operating system: Windows 7
  * Matplotlib version: 2.2.4 (same with 3.1.0)
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg (same with Qt5Agg/PySide2)
  * Python version: 3.6.5 32bit

installed with pip

**Comment by anntzer** [comment]
Looks like something wrong with Agg...  Likely we should investigate the lowest value Agg can draw properly and clip linewidths in the backend_agg graphicscontext?

**Comment by timhoffm** [comment]
Turns out, small linewidths require a lot of computation. When running the above code with different linewidths, the plot window appears, but is black for a certian time. During that time CPU usage is at 100%. Computation time goes up very strongly when reducing the linewidth:
~~~
lw=1e-4   # ~1s
lw=1e-5   # ~6s
lw=1e-6   # ~54s
~~~

Further notes:
- the delay is caused by antialiasing (setting `rcParams['lines.antialiased'] = False` results in immediate draw).
- resizing the plot also lags for small linewidths.
- with `lw=1e-3` the initial draw is not noticably slower, however, resizing already feels a bit laggy.
- for values `lw=1e-2` and smaller, no visible line is drawn at all.

### What to do?
From a practical point of view everything equal or smaller than `lw=1e-2` is identical (not drawn). I also don't think there are practical purposes for lines smaller than 1e-2 points. It should therefore be safe to clip `lw` to 1e-2.

**Comment by WeatherGod** [comment]
I wonder if this is also partly dependent upon the figure's dpi setting? If
so, then a straight-up clipping of 1e-2 is not sufficient as at high dpi,
it might still be visible, perhaps?

**Comment by timhoffm** [comment]
Even 1e-3 is not sufficient for our test cases. See #14560. Needs more investigation.

**Comment by tacaswell** [comment]
Why are thin lines computationally expensive?

**Comment by timhoffm** [comment]
It seems to have to do with antialiasing. But not an expert in the field, I neither know what type of antialiasing agg is using, nor how thin lines affect the performance of antialiasing.

**Comment by ImportanceOfBeingErnest** [comment]
> Why are thin lines computationally expensive?

I don't think thin lines are per se computationally more expensive than thick lines; it's just that **dashed** lines have more dashes the smaller their width. I.e. the number of dashes per unit length, and hence the number of points the renderer needs to draw, grows because there is a fixed dash-length to width ratio. 

![image](https://user-images.githubusercontent.com/23121882/60775409-192fc880-a123-11e9-9cbc-2fe2247aa5d2.png)

**Comment by timhoffm** [comment]
@ImportanceOfBeingErnest :+1: That explains it. Non-antialiased lines are clipped to a width >= 0.5
https://github.com/matplotlib/matplotlib/blob/db90685fe0522ed18c61848a5996b987a5d06a78/src/_backend_agg.h#L408
so they do not suffer from the increased number of points for small widths and thus do not have a performance issue.

**Comment by jklymak** [comment]
So can we just clip the dash length?

**Comment by tacaswell** [comment]
Sigh, the dash scaling is one of those rabbit holes with a :dragon_face: at the bottom.

We talked about clipping that scaling, but decided not to due to concerns about negatively affecting hi-dpi outputs.

attn @dopplershift as I recall you were the most vocal on not clipping the dash scaling in the run up to 2.0.

**Comment by timhoffm** [comment]
Adding an additional safety factor of 2 or 4 for hi-dpi should be enough, I mean how high can hi-dpi be compared to a standard resolution? If you are really concerned still, one could make that cutoff configurable, either in rcParams or in some global variable.  That way users with extreme needs would still be able to draw any dashing correctly.

**Comment by jklymak** [comment]
I agree.  How high can hi dpi go?  Are there really practical uses to 10,000 dpi?

**Comment by jklymak** [comment]
So `lw=1e-2` means you need to be running 7200 dpi to see that line.   I think its fair to clip at 1e-2 and let the world know that Matplotlib doesn’t support higher dpi than this.  Note that no fonts can be smaller than 1 point already due to the font rendering library limits, so its really hard to see why anyone would need a 1-point lettering and lines that are 100 times thinner than the font size.

**Comment by dopplershift** [comment]
So the original issue where I was arguing against clipping was #8032. The crux of my argument back then was that dots need to be dots, and we need to make sure we keep high DPI use cases in mind.

I'm not sure where lw=1e-4 comes into practice, but If we can solve this performance issue with even a rough clip at 1e-2 (and that indeed doesn't make a difference until you're higher than 7200 dpi), then I suppose that would be ok.

**Comment by jwuphysics** [comment]
I get a similar freeze when trying to input `dashes=(0, 0)`. I first encountered it in the Jupyter notebook environment and duplicated it in IPython. The window becomes unresponsive and my laptop fans kick in so I have to end the process.

I am using Python 3.7.3, matplotlib 3.1.1, Qt5Agg backend, and Linux Mint 19.1.

**Comment by oscargus** [comment]
> I get a similar freeze when trying to set `dashes=(0, 0)`

This part is fixed with #22569
